src/function/scalar/list/list_lambdas.cpp

```diff
@@ -0,0 +1,375 @@
+	// lhs lambda parameters, and to ensure that the lambda expression is not NULL
+	idx_t num_params = 0;
+	if (!arguments[arguments.size() - 1]->alias.empty()) {
+		num_params = stoi(arguments[arguments.size() - 1]->alias);
+	}
+	if (num_params != LAMBDA_PARAM_CNT) {
+		throw BinderException("Incorrect number of parameters in lambda function! " + bound_function.name +
+		                      " expects " + to_string(LAMBDA_PARAM_CNT) + " parameter(s).");
+	}
+
+	// remove the lambda function
+	auto lambda_expr = move(arguments.back());
+	arguments.pop_back();
+
+	if (arguments[0]->return_type.id() == LogicalTypeId::SQLNULL) {
+		bound_function.arguments.emplace_back(LogicalType::SQLNULL);
+		bound_function.return_type = LogicalType::SQLNULL;
+		return make_unique<VariableReturnBindData>(bound_function.return_type);
+	}
+
+	D_ASSERT(arguments[0]->return_type.id() == LogicalTypeId::LIST);
+
+	// push back the correct return types into the bound_function arguments
+	for (idx_t i = 0; i < arguments.size(); i++) {
+		bound_function.arguments.push_back(arguments[i]->return_type);
+	}
+
+	return make_unique<ListLambdaBindData>(bound_function.return_type, move(lambda_expr));
+}
+
+static unique_ptr<FunctionData> ListTransformBind(ClientContext &context, ScalarFunction &bound_function,
+                                                  vector<unique_ptr<Expression>> &arguments) {
+
+	// at least the list column and the lambda function
+	if (arguments.size() < 2) {
+		throw BinderException(bound_function.name + " takes two parameters (LIST, LAMBDA).");
+	}
+
+	bound_function.return_type = LogicalType::LIST(arguments[arguments.size() - 1]->return_type);
+	return ListLambdaBind<1>(context, bound_function, arguments);
+}
+
+static unique_ptr<FunctionData> ListFilterBind(ClientContext &context, ScalarFunction &bound_function,
+                                               vector<unique_ptr<Expression>> &arguments) {
+
+	// at least the list column and the lambda function
+	if (arguments.size() < 2) {
+		throw BinderException(bound_function.name + " takes two parameters (LIST, LAMBDA).");
+	}
+
+	bound_function.return_type = arguments[0]->return_type;
+	return ListLambdaBind<1>(context, bound_function, arguments);
+}
+
+void ListTransformFun::RegisterFunction(BuiltinFunctions &set) {
+
+	ScalarFunction fun("list_transform", {}, LogicalType::LIST(LogicalType::ANY), ListTransformFunction, false,
+	                   ListTransformBind, nullptr, nullptr);
+	fun.varargs = LogicalType::ANY;
+	set.AddFunction(fun);
+
+	fun.name = "array_transform";
+	set.AddFunction(fun);
+	fun.name = "list_apply";
+	set.AddFunction(fun);
+	fun.name = "array_apply";
+	set.AddFunction(fun);
+}
+
+void ListFilterFun::RegisterFunction(BuiltinFunctions &set) {
+
+	ScalarFunction fun("list_filter", {}, LogicalType::LIST(LogicalType::ANY), ListFilterFunction, false,
+	                   ListFilterBind, nullptr, nullptr);
+	fun.varargs = LogicalType::ANY;
```

Comment:
Why is this a varargs function? Shouldn't this be `LogicalType::LIST(LogicalType::ANY), LogicalType::LAMBDA`?

Reply:
When making the iteration of the lambda expression more generic, I moved it into the `BindLambdaFunction`. Here, I iterate and transform the children of the lambda expression and push them into the `children` vector.
```c++
auto bound_lambda_expr = move(children.back());
children.pop_back();
IterateLambdaExprChildren(children, list_child_type, bound_lambda_expr);
children.push_back(move(bound_lambda_expr));
```
After that, I call `ScalarFunction::BindScalarFunction(context, *func, move(children), error, function.is_operator);`. This means that the number of elements in `children` is not 2. If I don't make it a varargs function, then the binder complains.
```
Binder Error: No function matches the given name and argument types 'list_transform(INTEGER[], INTEGER, INTEGER)'. You might need to add explicit type casts.
	Candidate functions:
	list_transform() -> ANY[]
```
Is there a better solution for this?